shell/platform/common/cpp/client_wrapper/include/flutter/event_channel.h

```diff
@@ -0,0 +1,137 @@
+// binary before being transmitted back to Flutter. The MethodCodec used must be
+// compatible with the one used by the Flutter application. This can be achieved
+// by creating an EventChannel
+// ("https://docs.flutter.io/flutter/services/EventChannel-class.html")
+// counterpart of this channel on the Dart side.
+// The C++ type of stream configuration arguments, events, and error details are
+// Template, but only values supported by the specified MethodCodec can be used.
+template <typename T>
+class EventChannel {
+ public:
+  // Creates an instance that sends and receives event handler on the channel
+  // named |name|, encoded with |codec| and dispatched via |messenger|.
+  EventChannel(BinaryMessenger* messenger,
+               const std::string& name,
+               const MethodCodec<T>* codec)
+      : messenger_(messenger), name_(name), codec_(codec) {}
+  ~EventChannel() = default;
+
+  // Prevent copying.
+  EventChannel(EventChannel const&) = delete;
+  EventChannel& operator=(EventChannel const&) = delete;
+
+  // Registers a stream handler on this channel.
+  // If no handler has been registered, any incoming stream setup requests will
+  // be handled silently by providing an empty stream.
+  void SetStreamHandler(StreamHandler<T>* handler) const {
+    if (!handler) {
+      messenger_->SetMessageHandler(name_, nullptr);
+      return;
+    }
+
+    const MethodCodec<T>* codec = codec_;
+    const std::string channel_name = name_;
+    const BinaryMessenger* messenger = messenger_;
+    BinaryMessageHandler binary_handler = [handler, codec, channel_name,
+                                           messenger](
+                                              const uint8_t* message,
+                                              const size_t message_size,
+                                              BinaryReply reply) mutable {
+      constexpr char kOnListenMethod[] = "listen";
+      constexpr char kOnCancelMethod[] = "cancel";
+
+      std::unique_ptr<MethodCall<T>> method_call =
+          codec->DecodeMethodCall(message, message_size);
+      if (!method_call) {
+        std::cerr << "Unable to construct method call from message on channel: "
+                  << channel_name << std::endl;
+        return;
+      }
+
+      const std::string& method = method_call->method_name();
+      if (method.compare(kOnListenMethod) == 0) {
+        auto sink = std::make_unique<EventSinkImplementation>(
+            messenger, channel_name, codec);
+        handler->onListen(method_call->arguments(), std::move(sink));
+
+        auto result = codec->EncodeSuccessEnvelope();
+        reply(result->data(), result->size());
+      } else if (method.compare(kOnCancelMethod) == 0) {
+        handler->onCancel(method_call->arguments());
+
+        auto result = codec->EncodeSuccessEnvelope();
+        reply(result->data(), result->size());
+      } else {
+        std::cerr
```

Comment:
Logging unhandled methods should be handled on the Flutter side already, so this shouldn't be needed.